File: tire_data.py
import sqlite3
from datetime import datetime

def connect_db():
    conn = sqlite3.connect('tires.db')
    return conn

def insert_tire_sizes():
    # Common tire sizes with their specifications
    tire_sizes = [
        # Passenger Car Sizes
        (205, 55, 16, '91', 'V', 'R', '205/55R16'),
        (215, 55, 17, '94', 'V', 'R', '215/55R17'),
        (225, 45, 17, '91', 'Y', 'R', '225/45R17'),
        (235, 45, 18, '94', 'Y', 'R', '235/45R18'),
        (245, 40, 18, '93', 'Y', 'R', '245/40R18'),
        (255, 35, 19, '96', 'Y', 'R', '255/35R19'),
        
        # SUV Sizes
        (225, 65, 17, '102', 'H', 'R', '225/65R17'),
        (235, 65, 17, '104', 'H', 'R', '235/65R17'),
        (245, 65, 17, '107', 'H', 'R', '245/65R17'),
        (255, 65, 17, '109', 'H', 'R', '255/65R17'),
        (265, 65, 17, '112', 'H', 'R', '265/65R17'),
        
        # Light Truck Sizes
        (265, 70, 17, '115', 'S', 'R', '265/70R17'),
        (275, 70, 17, '117', 'S', 'R', '275/70R17'),
        (285, 70, 17, '121', 'S', 'R', '285/70R17'),
        (295, 70, 17, '121', 'S', 'R', '295/70R17'),
        (305, 70, 17, '121', 'S', 'R', '305/70R17')
    ]
    
    conn = connect_db()
    cursor = conn.cursor()
    
    for size in tire_sizes:
        cursor.execute('''
            INSERT OR IGNORE INTO tire_size 
            (width, aspect_ratio, rim_size, load_rating, speed_rating, construction, full_size)
            VALUES (?, ?, ?, ?, ?, ?, ?)
        ''', size)
    
    conn.commit()
    conn.close()

def insert_tire_models():
    # Tire models with their specifications
    tire_models = [
        # Michelin Models
        ("Michelin", "Pilot Sport 4S", "245/40R18", 220.0, "Ultra High Performance", 9, 30000, 9, 9, 7, True, "Premium summer tire", 100, "Run-flat, Low rolling resistance"),
        ("Michelin", "Primacy MXM4", "215/55R17", 180.0, "All-Season", 8, 45000, 8, 8, 8, False, "All-season touring tire", 150, "All-season traction, Comfort ride"),
        
        # Continental Models
        ("Continental", "ExtremeContact DWS06", "225/45R17", 190.0, "All-Season", 9, 40000, 9, 9, 7, False, "All-season performance tire", 120, "Dry, Wet, Snow traction"),
        ("Continental", "PureContact LS", "205/55R16", 160.0, "All-Season", 8, 70000, 8, 8, 8, False, "All-season touring tire", 200, "Long tread life, Fuel efficient"),
        
        # Bridgestone Models
        ("Bridgestone", "Potenza RE980AS", "235/45R18", 200.0, "All-Season", 9, 50000, 9, 9, 7, False, "All-season performance tire", 180, "All-season traction, Sport handling"),
        ("Bridgestone", "Dueler H/L Alenza", "265/65R17", 180.0, "All-Season", 8, 65000, 8, 8, 8, False, "SUV touring tire", 150, "SUV optimized, All-season traction"),
        
        # Goodyear Models
        ("Goodyear", "Eagle F1 Asymmetric", "255/35R19", 210.0, "Ultra High Performance", 9, 30000, 9, 9, 6, True, "Summer performance tire", 100, "Sport handling, Wet traction"),
        ("Goodyear", "Assurance WeatherReady", "225/60R16", 170.0, "All-Season", 8, 60000, 8, 8, 8, False, "All-season touring tire", 200, "All-weather traction, Comfort ride"),
        
        # Pirelli Models
        ("Pirelli", "P Zero", "245/40R18", 230.0, "Ultra High Performance", 9, 30000, 9, 9, 6, True, "Summer performance tire", 80, "Sport handling, Wet traction"),
        ("Pirelli", "Cinturato P7", "225/45R17", 190.0, "All-Season", 8, 50000, 8, 8, 8, False, "All-season touring tire", 150, "All-season traction, Fuel efficient")
    ]
    
    conn = connect_db()
    cursor = conn.cursor()
    
    for model in tire_models:
        # Get size_id
        cursor.execute('SELECT id FROM tire_size WHERE full_size = ?', (model[2],))
        size_id = cursor.fetchone()
        
        if size_id:
            # Calculate prices
            base_price = model[3]
            offer_price = base_price * 0.9
            bulk_price = base_price * 0.85
            
            cursor.execute('''
                INSERT INTO tire_product 
                (brand, model, size_id, base_price, offer_price, bulk_price, category, 
                performance_rating, tread_life, wet_performance, dry_performance, 
                noise_level, run_flat, description, stock, features, created_at)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
            ''', (model[0], model[1], size_id[0], base_price, offer_price, bulk_price,
                 model[4], model[5], model[6], model[7], model[8], model[9],
                 model[10], model[11], model[12], model[13], datetime.now()))
    
    conn.commit()
    conn.close()

def insert_vehicle_compatibility():
    # Sample vehicle-tire compatibility data
    compatibility_data = [
        # BMW Models
        ("BMW", "3 Series", 2020, "225/45R17", True),
        ("BMW", "5 Series", 2020, "245/40R18", True),
        ("BMW", "X5", 2020, "265/65R17", True),
        
        # Mercedes Models
        ("Mercedes", "C-Class", 2020, "225/45R17", True),
        ("Mercedes", "E-Class", 2020, "245/40R18", True),
        ("Mercedes", "GLE", 2020, "265/65R17", True),
        
        # Audi Models
        ("Audi", "A4", 2020, "225/45R17", True),
        ("Audi", "A6", 2020, "245/40R18", True),
        ("Audi", "Q5", 2020, "235/65R17", True),
        
        # Toyota Models
        ("Toyota", "Camry", 2020, "215/55R17", True),
        ("Toyota", "RAV4", 2020, "225/65R17", True),
        ("Toyota", "Highlander", 2020, "245/65R17", True)
    ]
    
    conn = connect_db()
    cursor = conn.cursor()
    
    for comp in compatibility_data:
        # Insert vehicle if not exists
        cursor.execute('''
            INSERT OR IGNORE INTO vehicle (make, model, year, tire_size, created_at)
            VALUES (?, ?, ?, ?, ?)
        ''', (comp[0], comp[1], comp[2], comp[3], datetime.now()))
        
        # Get vehicle_id
        cursor.execute('SELECT id FROM vehicle WHERE make = ? AND model = ? AND year = ?',
                      (comp[0], comp[1], comp[2]))
        vehicle_id = cursor.fetchone()
        
        if vehicle_id:
            # Get tire_id
            cursor.execute('''
                SELECT tp.id 
                FROM tire_product tp
                JOIN tire_size ts ON tp.size_id = ts.id
                WHERE ts.full_size = ?
            ''', (comp[3],))
            tire_ids = cursor.fetchall()
            
            # Create compatibility relationships
            for tire_id in tire_ids:
                cursor.execute('''
                    INSERT OR IGNORE INTO tire_vehicle_compatibility 
                    (tire_id, vehicle_id, is_original_equipment)
                    VALUES (?, ?, ?)
                ''', (tire_id[0], vehicle_id[0], comp[4]))
    
    conn.commit()
    conn.close()

def calculate_savings(tire_id, quantity):
    """Calculate total savings for different purchase quantities"""
    conn = connect_db()
    cursor = conn.cursor()
    
    cursor.execute('''
        SELECT base_price, offer_price, bulk_price
        FROM tire_product
        WHERE id = ?
    ''', (tire_id,))
    
    prices = cursor.fetchone()
    conn.close()
    
    if not prices:
        return None
    
    base_price, offer_price, bulk_price = prices
    
    if quantity >= 4:
        total_savings = (base_price - bulk_price) * quantity
        discount_type = "Bulk (15%)"
    else:
        total_savings = (base_price - offer_price) * quantity
        discount_type = "Regular (10%)"
    
    return {
        "original_total": base_price * quantity,
        "discounted_total": (bulk_price if quantity >= 4 else offer_price) * quantity,
        "total_savings": total_savings,
        "discount_type": discount_type
    }

if __name__ == "__main__":
    print("Inserting tire sizes...")
    insert_tire_sizes()
    
    print("Inserting tire models...")
    insert_tire_models()
    
    print("Inserting vehicle compatibility...")
    insert_vehicle_compatibility()
    
    print("Data insertion completed successfully") 

File: data/tireData.ts
export interface Tire {
  id: string;
  brand: string;
  model: string;
  size: string;
  base_price: number;
  description: string;
  image_url: string;
  specifications: {
    loadIndex: string;
    speedRating: string;
    season: 'Summer' | 'Winter' | 'All-Season';
    runFlat: boolean;
  };
}

// Tire recommendations for specific car models
const carTireRecommendations: { [key: string]: { size: string; recommended: string[] } } = {
  // BMW Models
  '3-Series': {
    size: "225/45R17",
    recommended: ['Michelin', 'Bridgestone', 'Continental'] // Sport sedan focus
  },
  '5-Series': {
    size: "245/45R18",
    recommended: ['Michelin', 'Continental', 'Pirelli'] // Luxury sedan focus
  },
  'X5': {
    size: "255/50R19",
    recommended: ['Bridgestone', 'Pirelli', 'Goodyear'] // SUV focus
  },
  
  // Audi Models
  'A4': {
    size: "225/50R17",
    recommended: ['Continental', 'Michelin', 'Goodyear'] // Sport sedan focus
  },
  'A6': {
    size: "245/45R18",
    recommended: ['Pirelli', 'Michelin', 'Continental'] // Luxury sedan focus
  },
  'Q5': {
    size: "235/60R18",
    recommended: ['Goodyear', 'Bridgestone', 'Continental'] // SUV focus
  },
  
  // Mercedes Models
  'C-Class': {
    size: "225/45R17",
    recommended: ['Continental', 'Michelin', 'Pirelli'] // Sport luxury focus
  },
  'E-Class': {
    size: "245/45R18",
    recommended: ['Michelin', 'Pirelli', 'Continental'] // Luxury focus
  },
  'GLC': {
    size: "235/60R18",
    recommended: ['Bridgestone', 'Goodyear', 'Continental'] // SUV focus
  }
};

// Helper function to get tire image based on brand
function getTireImage(brand: string): string {
  switch (brand.toLowerCase()) {
    case 'michelin':
      return '/data/Tyre.webp';
    case 'continental':
      return '/data/Tyre2.webp';
    case 'bridgestone':
      return '/data/Tyre3.webp';
    case 'goodyear':
      return '/data/Tyre4.webp';
    case 'pirelli':
      return '/data/Tyre5.webp';
    default:
      return '/data/Tyre.webp';
  }
}

// Tire model specifications
const tireModels: { [key: string]: { model: string; description: string; base_price: number; specs: any } } = {
  'Michelin': {
    model: "Pilot Sport 4S",
    description: "High-performance summer tire with excellent grip and handling",
    base_price: 199.99,
    specs: {
      loadIndex: "91",
      speedRating: "Y",
      season: "Summer",
      runFlat: true
    }
  },
  'Continental': {
    model: "ExtremeContact DWS06",
    description: "All-season tire with excellent wet and dry performance",
    base_price: 179.99,
    specs: {
      loadIndex: "91",
      speedRating: "Y",
      season: "All-Season",
      runFlat: false
    }
  },
  'Bridgestone': {
    model: "Potenza Sport",
    description: "Ultra-high performance tire with maximum grip and control",
    base_price: 189.99,
    specs: {
      loadIndex: "91",
      speedRating: "Y",
      season: "Summer",
      runFlat: true
    }
  },
  'Goodyear': {
    model: "Eagle F1 Asymmetric 6",
    description: "Premium performance tire with superior wet handling",
    base_price: 185.99,
    specs: {
      loadIndex: "91",
      speedRating: "Y",
      season: "Summer",
      runFlat: false
    }
  },
  'Pirelli': {
    model: "P Zero",
    description: "Ultra-high performance tire for luxury and sports cars",
    base_price: 209.99,
    specs: {
      loadIndex: "91",
      speedRating: "Y",
      season: "Summer",
      runFlat: true
    }
  }
};

// Helper function to generate tire data for a specific model and size
function generateTireData(modelKey: string, size: string): Tire[] {
  console.log('Generating tire data for model:', modelKey); // Debug log
  
  // Remove any potential underscores from the model key
  const carModel = modelKey.replace('_', '-');
  console.log('Looking up recommendations for:', carModel); // Debug log
  
  const recommendations = carTireRecommendations[carModel]?.recommended || Object.keys(tireModels);
  console.log('Found recommendations:', recommendations); // Debug log
  
  const tires = recommendations.map(brand => {
    const tireModel = tireModels[brand];
    return {
      id: `${brand.toLowerCase()}-${modelKey}`,
      brand: brand,
      model: tireModel.model,
      size: size,
      base_price: tireModel.base_price,
      description: tireModel.description,
      image_url: getTireImage(brand),
      specifications: tireModel.specs
    };
  });
  
  console.log(`Generated ${tires.length} tires for ${modelKey}`); // Debug log
  return tires;
}

// Generate database entries for all years
const years = Array.from({ length: 10 }, (_, i) => (2024 - i).toString());
const tireDatabase: { [key: string]: Tire[] } = {};

// BMW Models
years.forEach(year => {
  tireDatabase[`BMW-3-Series-${year}`] = generateTireData('3-Series', carTireRecommendations['3-Series'].size);
  tireDatabase[`BMW-5-Series-${year}`] = generateTireData('5-Series', carTireRecommendations['5-Series'].size);
  tireDatabase[`BMW-X5-${year}`] = generateTireData('X5', carTireRecommendations['X5'].size);
});

// Audi Models
years.forEach(year => {
  tireDatabase[`Audi-A4-${year}`] = generateTireData('A4', carTireRecommendations['A4'].size);
  tireDatabase[`Audi-A6-${year}`] = generateTireData('A6', carTireRecommendations['A6'].size);
  tireDatabase[`Audi-Q5-${year}`] = generateTireData('Q5', carTireRecommendations['Q5'].size);
});

// Mercedes Models
years.forEach(year => {
  tireDatabase[`Mercedes-C-Class-${year}`] = generateTireData('C-Class', carTireRecommendations['C-Class'].size);
  tireDatabase[`Mercedes-E-Class-${year}`] = generateTireData('E-Class', carTireRecommendations['E-Class'].size);
  tireDatabase[`Mercedes-GLC-${year}`] = generateTireData('GLC', carTireRecommendations['GLC'].size);
});

export function getTiresForVehicle(make: string, model: string, year: string): Tire[] {
  // Format the key to match the database
  const key = `${make}-${model}-${year}`;
  console.log('Searching for key:', key); // Debug log
  
  // Get the tires from the database
  const tires = tireDatabase[key] || [];
  console.log('Found tires:', tires.length); // Debug log
  
  // If no tires found, return all available tires for that size
  if (tires.length === 0) {
    console.log('No specific recommendations found, returning all tire options');
    const defaultSize = "225/45R17"; // Default size if none found
    return Object.keys(tireModels).map(brand => {
      const tireModel = tireModels[brand];
      return {
        id: `${brand.toLowerCase()}-default`,
        brand: brand,
        model: tireModel.model,
        size: defaultSize,
        base_price: tireModel.base_price,
        description: tireModel.description,
        image_url: getTireImage(brand),
        specifications: tireModel.specs
      };
    });
  }
  
  return tires;
}

export const tireWidths = ["185", "195", "205", "215", "225", "235", "245", "255", "265", "275", "285", "295", "305", "315", "325", "335", "345", "355"];
export const tireProfiles = ["25", "30", "35", "40", "45", "50", "55", "60", "65", "70", "75", "80"];
export const rimSizes = ["14", "15", "16", "17", "18", "19", "20", "21", "22", "23", "24"]; 